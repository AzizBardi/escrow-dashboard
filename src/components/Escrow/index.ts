export { EscrowContainer as Escrow } from './Escrow';
