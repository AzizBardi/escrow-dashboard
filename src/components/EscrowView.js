import { Divider, Typography } from "@mui/material";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";

export default function Escrow(props) {
  const {
    status,
    address,
    repOracleAddr,
    repOracleStake,
    recOracleAddr,
    recOracleStake,
    balance,
    manifestUrl,
    finalResultsUrl,
  } = props;

  return (
    <Card variant="outlined">
      <CardContent>
          <CardTextBlock title="Address" value={address} />
          <CardTextBlock title="Status" value={status} />
          <CardTextBlock title="Reputation Oracle" value={repOracleAddr} />
          <CardTextBlock title="Reputation Oracle stake(%)" value={repOracleStake} />
          <CardTextBlock title="Recording Oracle" value={recOracleAddr} />
          <CardTextBlock title="Recording Oracle stake(%)" value={recOracleStake} />
          <CardTextBlock title="Balance(HMT)" value={balance} />
          <CardTextBlock title="Manifest" value={manifestUrl} />
          <CardTextBlock title="Final Results" value={finalResultsUrl} />
      </CardContent>
    </Card>
  );
}

function CardTextBlock({ title, value }) {
  return (
    <>
      <Divider textAlign="center" sx={{mt: 1}}>
        <Typography variant="body2" color="text.secondary">
          {title}
        </Typography>
      </Divider>
      <Typography variant="body2" textAlign="center">
        {value}
      </Typography>
    </>
  );
}
