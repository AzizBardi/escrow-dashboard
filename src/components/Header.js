import { useRecoilState, useSetRecoilState } from "recoil";
import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Toolbar from "@mui/material/Toolbar";

import {ReactComponent as MetamaskLogo} from '../metamask.svg';
import { walletState } from "../store/wallet";
import { isPolygon } from "../constants";
import { enableWeb3, switchToPolygon, getWeb3 } from "../WalletProvider";

function Header() {
  const [wallet] = useRecoilState(walletState);
  const setWallet = useSetRecoilState(walletState);

  const onConnect = async () => {
    const web3 = getWeb3();
    await enableWeb3();

    window.ethereum.on('chainChanged', (chainChanged) => {
      setWallet(oldWallet => ({ ...oldWallet, networkChainId: chainChanged }));
    });

    window.ethereum.on('accountsChanged', (accounts) => {
      if (accounts.length > 0) {
        setWallet(oldWallet => ({ ...oldWallet, address: accounts[0]}));
      } else {
        setWallet({ networkChainId: '', address: '', isConnected: false})
      }
    });

    const networkChainId = await window.ethereum.request({
      method: "eth_chainId",
    });
    const account = (await web3.eth.getAccounts())[0];
    setWallet((oldWallet) => ({
      ...oldWallet,
      address: account,
      networkChainId,
      isConnected: true,
    }));
  };

  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar position="static">
        <Toolbar>
          <Typography
            variant="h6"
            component="div"
            color="white"
            sx={{
              flexGrow: 1,
              paddingLeft: 2,
              paddingBottom: 2,
              paddingTop: 2,
            }}
          >
            Escrow scanner
          </Typography>
          <Button
            variant="contained"
            startIcon={<MetamaskLogo />}
            color="info"
            disabled={wallet.isConnected}
            onClick={onConnect}
          >
            Connect
          </Button>
          <Button
            variant="contained"
            color="info"
            disabled={!wallet.isConnected || isPolygon(wallet.networkChainId)}
            onClick={switchToPolygon}
            sx={{ marginLeft: "4px" }}
          >
            Switch To Polygon
          </Button>
        </Toolbar>
      </AppBar>
    </Box>
  );
}

export default Header;
