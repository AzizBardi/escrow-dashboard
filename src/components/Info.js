import { useRecoilState } from "recoil";
import Typography from "@mui/material/Typography";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import Divider from "@mui/material/Divider";

import { walletState } from "../store/wallet";
import { escrowAddressState } from "../store/escrow";
import { chainIdToTitle, chainIdToScannerUrl } from "../constants";

export default function Info() {
  const [wallet] = useRecoilState(walletState);
  const [escrowAddr] = useRecoilState(escrowAddressState);
  const networkScannerUrl = wallet.networkChainId && chainIdToScannerUrl[wallet.networkChainId];
  const escrowUrl = `${networkScannerUrl}/address/${escrowAddr}`;

  return (
    <Card variant="outlined" sx={{ width: "450px" }}>
      <CardContent>
        <Divider textAlign="center">
          <Typography color="text.secondary" variant="body2">
            Network
          </Typography>
        </Divider>
        <Typography variant="body2" textAlign="center">
          {chainIdToTitle[wallet.networkChainId]}
        </Typography>
        <Divider textAlign="center" sx={{ mt: 1 }}>
          <Typography color="text.secondary" variant="body2">
            Escrow
          </Typography>
        </Divider>
        <Typography variant="body2" textAlign="center">
          {escrowAddr && <a href={escrowUrl} rel="noopener noreferrer" target="_blank"> {escrowAddr}</a>}
        </Typography>
      </CardContent>
    </Card>
  );
}
