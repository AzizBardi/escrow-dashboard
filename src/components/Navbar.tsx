import * as React from 'react';
import { Box, Link, Container, Menu, Toolbar } from '@mui/material';
import { NavLink } from 'react-router-dom';
import { routes } from '../routes';

const Navbar: React.FC = (): React.ReactElement => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  // const handleOpenNavMenu = (event: React.MouseEvent<HTMLButtonElement>) => {
  //   setAnchorEl(event.currentTarget);
  // };

  const handleCloseNavMenu = () => {
    setAnchorEl(null);
  };

  return (
    <Box
      sx={{
        backgroundColor: 'primary.dark',
      }}
    >
      <Container maxWidth="xl">
        <Toolbar disableGutters>
          <Box sx={{ display: { xs: 'flex', md: 'none' } }}>
            <Menu
              id="menu-appbar"
              anchorEl={anchorEl}
              anchorOrigin={{
                vertical: 'bottom',
                horizontal: 'left',
              }}
              keepMounted
              transformOrigin={{
                vertical: 'top',
                horizontal: 'left',
              }}
              open={Boolean(anchorEl)}
              onClose={handleCloseNavMenu}
              sx={{
                display: { xs: 'block', md: 'none' },
              }}
            />
          </Box>
          <Box sx={{ display: { xs: 'none', md: 'flex' } }}>
            <Box
              sx={{
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'flex-start',
                alignItems: 'center',
              }}
            >
              {routes.map((page) => (
                <Link
                  key={page.key}
                  component={NavLink}
                  to={page.path}
                  color="info.main"
                  underline="none"
                  variant="button"
                  sx={{ fontSize: 'large', marginLeft: '2rem' }}
                >
                  {page.title}
                </Link>
              ))}
            </Box>
          </Box>
        </Toolbar>
      </Container>
    </Box>
  );
};

export default Navbar;
