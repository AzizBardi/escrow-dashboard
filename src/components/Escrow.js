import {useEffect, useState} from 'react';
import { useRecoilState, useRecoilValue } from 'recoil';

import {escrowAddressState} from '../store/escrow';
import {isConnectedSelector} from '../store/wallet';
import { getWeb3 } from '../WalletProvider';
import EscrowView from './EscrowView';
import EscrowABI from '../contracts/EscrowAbi.json';

const statusesMap = ['Launched', 'Pending', 'Partial', 'Paid', 'Complete', 'Cancelled'];

export default function EscrowContainer() {
    const [escrowAddr] = useRecoilState(escrowAddressState);
    const isConnected = useRecoilValue(isConnectedSelector);
    const [escrowStatus, setEscrowStatus] = useState('');

    const [reputationOracle, setReputationOracle] = useState('');
    const [reputationOracleStake, setReputationOracleStake] = useState('');
  
    const [recordingOracle, setRecordingOracle] = useState('');
    const [recordingOracleStake, setRecordingOracleStake] = useState('');
  
    const [manifestUrl, setManifestUrl] = useState('');
    const [finalResultsUrl, setFinalResultsUrl] = useState('');
    const [balance, setBalance] = useState('');
  
  

    useEffect(() => {
        if (!isConnected || !escrowAddr) {
            return null;
        }
        async function setupEscrow() {
            const web3 = getWeb3();
            const Escrow = new web3.eth.Contract(EscrowABI, escrowAddr);
            const escrowSt = await Escrow.methods.status().call();
            setEscrowStatus(statusesMap[escrowSt]);
        
            const recOracleAddr = await Escrow.methods.recordingOracle().call();
            setRecordingOracle(recOracleAddr);
        
            const recOracleStake = await Escrow.methods.recordingOracleStake().call();
            setRecordingOracleStake(recOracleStake);
        
            const repOracleAddr = await Escrow.methods.reputationOracle().call();
            setReputationOracle(repOracleAddr);
        
            const repOracleStake = await Escrow.methods.reputationOracleStake().call();
            setReputationOracleStake(repOracleStake);
        
            const finalResults = await Escrow.methods.finalResultsUrl().call();
            setFinalResultsUrl(finalResults);
        
            const manifest = await Escrow.methods.manifestUrl().call();
            setManifestUrl(manifest);
        
            const balance = await Escrow.methods.getBalance().call();
            setBalance(web3.utils.fromWei(balance, 'ether'));
        
        }

        setupEscrow();

    }, [escrowAddr, isConnected]);

    return (
        <EscrowView
          status={escrowStatus}
          address={escrowAddr}
          repOracleAddr={reputationOracle}
          recOracleAddr={recordingOracle}
          repOracleStake={reputationOracleStake}
          recOracleStake={recordingOracleStake}
          balance={balance}
          manifestUrl={manifestUrl}
          finalResultsUrl={finalResultsUrl}

        />
    )
}
