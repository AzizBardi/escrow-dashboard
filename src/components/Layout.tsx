import { Box } from '@mui/material';
import * as React from 'react';
import Header from './Header';

interface ILayout {
  children: React.ReactNode;
}

const Layout: React.FC<ILayout> = ({ children }): React.ReactElement => (
  <Box
    sx={{
      marginTop: '120px',
    }}
  >
    <Header />
    {children}
  </Box>
);

export default Layout;
