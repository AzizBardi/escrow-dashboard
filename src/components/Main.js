import {useState} from 'react';
import Box from "@mui/material/Box";

import Header from "./Header";
import Escrow from "./Escrow";
import Search from "./Search";
import Footer from "./Footer";

function Main() {
  const [escrowFactory, setEscrowFactory] = useState('0x45eBc3eAE6DA485097054ae10BA1A0f8e8c7f794');

  return (
    <Box>
      <Header />
      <Box sx={{ display: "flex", justifyContent: "center", alignItems: 'center'}}>
        <Box sx={{ display: "flex", flexDirection: "column", justifyContent: "center" }}>
          <Search onSetEscrow={setEscrowFactory}></Search>
          <Box sx={{ display: "flex", flexDirection: "row", justifyContent: "center", mt: 2 }}>
            <Escrow address={escrowFactory} />
          </Box>
          <Footer />
        </Box>
      </Box>
      <Box>

      </Box>

    </Box>
  );
}

export default Main;
