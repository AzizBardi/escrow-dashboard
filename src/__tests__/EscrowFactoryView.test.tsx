import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { EscrowFactoryView } from 'src/components/Escrow';

const mock = {
  count: 10,
  address: '63fea823365f1c81fad234abdf5a1f43',
  latestEscrow: 'eaac4d45c3c41f449cf7c94622afacbc',
  eventsUrl: 'test.url',
  scanner: 'test-scanner',
  pendingEventCount: 1,
  bulkTransferEventCount: 1,
  intermediateStorageEventCount: 1,
};

describe('when rendered CardTextBlock component', () => {
  it('should render passed prop `count`', () => {
    render(
      <EscrowFactoryView
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
        bulkTransferEventCount={mock.bulkTransferEventCount}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    );
    expect(screen.getByText(mock.count)).toBeInTheDocument();
  });

  it('should render passed prop `address`', () => {
    render(
      <EscrowFactoryView
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
        bulkTransferEventCount={mock.bulkTransferEventCount}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    );
    expect(screen.findByLabelText(mock.address)).toBeTruthy();
  });

  it('should render passed prop `latestEscrow`', () => {
    render(
      <EscrowFactoryView
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
        bulkTransferEventCount={mock.bulkTransferEventCount}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    );
    expect(screen.findByLabelText(mock.latestEscrow)).toBeTruthy();
  });

  it('should render passed prop `eventsUrl`', () => {
    render(
      <EscrowFactoryView
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
        bulkTransferEventCount={mock.bulkTransferEventCount}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    );
    expect(screen.findByLabelText(mock.eventsUrl)).toBeTruthy();
  });

  it('should render passed prop `scanner`', () => {
    render(
      <EscrowFactoryView
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
        bulkTransferEventCount={mock.bulkTransferEventCount}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    );
    expect(screen.findByLabelText(mock.scanner)).toBeTruthy();
  });
});

it('EscrowFactoryView component renders correctly, corresponds to the snapshot', () => {
  const tree = renderer
    .create(
      <EscrowFactoryView
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
        bulkTransferEventCount={mock.bulkTransferEventCount}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
