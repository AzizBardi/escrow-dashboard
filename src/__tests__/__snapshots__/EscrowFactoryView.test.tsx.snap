// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EscrowFactoryView component renders correctly, corresponds to the snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Escrow Factory Address
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      63fea823365f1c81fad234abdf5a1f43
    </p>
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Latest Escrow
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      eaac4d45c3c41f449cf7c94622afacbc
    </p>
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Amount Of Escrows
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      10
    </p>
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          All Escrows Pending Events
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      1
    </p>
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          All Escrows BulkTransfer Events
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      1
    </p>
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          All Escrows IntermediateStorage Events
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      1
    </p>
    <div
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
      role="separator"
    >
      <span
        className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Total Number Of Escrows Events
        </p>
      </span>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
    >
      3
    </p>
    <div
      className="MuiBox-root css-xi606m"
    >
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-16qoh2c-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            All Launched Escrows
          </p>
        </span>
      </div>
      <a
        className="MuiTypography-root MuiTypography-inherit MuiTypography-alignCenter MuiLink-root MuiLink-underlineAlways css-1qm1xlu-MuiTypography-root-MuiLink-root"
        href="test.url"
        onBlur={[Function]}
        onFocus={[Function]}
        rel="noreferrer"
        target="_blank"
      >
        test-scanner
      </a>
    </div>
  </div>
</div>
`;
